components/collections/codepointtrie_builder/cpp/list_to_ucptrie.cpp

```diff
@@ -2,111 +2,95 @@
+#include "unicode/umutablecptrie.h"
+#include "writesrc.h"
 
 /**
  * list_to_ucptrie: converts a stream of integers to a UCPTrie
  *
- * - standard input should be a stream of integers, each integer representing the value for
-     consecutive code points, starting from code point 0.
+ * - standard input should be a stream of integers, each integer representing
+ the value for consecutive code points, starting from code point 0.
  * - standard output is the build UCPTrie as a TOML file
  */
-int main(int argc, char const *argv[]) {
-    if (argc != 5) {
-        std::cerr << "Takes 4 positional arguments: default value, error value, trie type, and value width" << std::endl;
-        return 1;
-    }
-
-    int32_t defaultValue = atoi(argv[1]);
-    int32_t errorValue = atoi(argv[2]);
 
-    UCPTrieType trieType;
-    std::string trieTypeStr(argv[3]);
-    if (trieTypeStr == "small") {
-        trieType = UCPTRIE_TYPE_SMALL;
-    } else if (trieTypeStr == "fast") {
-        trieType = UCPTRIE_TYPE_FAST;
-    } else {
-        std::cerr << "Expected 'small' or 'fast' for 3rd argument" << std::endl;
-        return 1;
-    }
+constexpr char TRIE_SMALL[] = "small";
+constexpr char TRIE_FAST[] = "fast";
+extern "C" {
+int construct_ucptrie(const int32_t defaultValue, const int32_t errorValue,
+                      const char* trieTypeStr, const int32_t valueWidthInt,
+                      const uint32_t* values, const uint32_t n) {
+  UCPTrieType trieType;
+  if (strcmp(trieTypeStr, TRIE_SMALL) == 0) {
+    trieType = UCPTRIE_TYPE_SMALL;
+  } else if (strcmp(trieTypeStr, TRIE_FAST) == 0) {
+    trieType = UCPTRIE_TYPE_FAST;
+  } else {
+    fprintf(stderr, "Expected 'small' or 'fast' for 3rd argument. Got %s",
+            trieTypeStr);
+    return 1;
+  }
 
-    UCPTrieValueWidth valueWidth;
-    int32_t valueWidthInt = atoi(argv[4]);
-    if (valueWidthInt == 8) {
-        valueWidth = UCPTRIE_VALUE_BITS_8;
-    } else if (valueWidthInt == 16) {
-        valueWidth = UCPTRIE_VALUE_BITS_16;
-    } else if (valueWidthInt == 32) {
-        valueWidth = UCPTRIE_VALUE_BITS_32;
-    } else {
-        std::cerr << "Expected '8', '16', or '32' for 4th argument" << std::endl;
-        return 1;
-    }
+  UCPTrieValueWidth valueWidth;
+  if (valueWidthInt == 8) {
+    valueWidth = UCPTRIE_VALUE_BITS_8;
+  } else if (valueWidthInt == 16) {
+    valueWidth = UCPTRIE_VALUE_BITS_16;
+  } else if (valueWidthInt == 32) {
+    valueWidth = UCPTRIE_VALUE_BITS_32;
+  } else {
+    fprintf(stderr, "Expected '8', '16', or '32' for 4th argument. Got %d",
+            valueWidthInt);
+    return 1;
+  }
 
-    icu::ErrorCode status;
+  icu::ErrorCode status;
 
-    icu::LocalUMutableCPTriePointer builder(umutablecptrie_open(
-        defaultValue,
-        errorValue,
-        status));
-    if (status.isFailure()) {
-        std::cerr << status.errorName() << std::endl;
-        return 1;
-    }
+  icu::LocalUMutableCPTriePointer builder(
```

Comment:
nit: this is the first time I'm reading this code, and the C++ parenthesis initialization always trips me up. Consider using `auto builder = umutablecptrie_open(defaultValue, errorValue, status);` to make this more readable to Rust devs who don't know C++.

Reply:
Done.